### ISSUE TITLE
Crash on torrent load

### ISSUE THREAD

**Issue by ngosang** [arvidn/libtorrent#1528]
qBittorrent version: v3.3.5
Libtorrent version: 1.0.9.0
Qt version: 5.6.1
Boost version: 1.61.0
OS version: Windows 10 10.0.10586 x86_64

Could you test the torrent file provided? https://github.com/qbittorrent/qBittorrent/issues/5480#issuecomment-230411200
Thank you

**Comment by ngosang** [comment]
Reported in https://github.com/qbittorrent/qBittorrent/issues/5480

**Comment by sledgehammer999** [comment]
It crashes with latest RC_1_0.

Relevant backtrace:

```
#  8 qbittorrent.exe      0x000000013fbd1710 memcpy()[ f:\dd\vctools\crt\vcruntime\src\string\amd64\memcpy.asm : 138 ]
#  9 qbittorrent.exe      0x000000013f313dac libtorrent::torrent_info::hash_for_piece(index)[ g:\qbittorrent\libtorrent\include\libtorrent\torrent_info.hpp : 561 ]
# 10 qbittorrent.exe      0x000000013f32764a libtorrent::create_torrent::create_torrent(ti)[ g:\qbittorrent\libtorrent\src\create_torrent.cpp : 360 ]
# 11 qbittorrent.exe      0x000000013f1d67f1 BitTorrent::TorrentHandle::saveTorrentFile(path)[ base\bittorrent\torrenthandle.cpp : 1315 ]
# 12 qbittorrent.exe      0x000000013f1bf622 BitTorrent::Session::createTorrentHandle(nativeHandle)[ base\bittorrent\session.cpp : 3295 ]
```
@ngosang I think that's why I didn't report. The backtrace needs a little bit investigation on our part. But now I see it still a libtorrent problem.

RC_1_1 fails much more gracefully. When adding the torrent we always check if the obtained torrent_info is valid. RC_1_1 tells that it isn't with error: `incorrect number of piece hashes in torrent file`
While RC_1_0 tells that this is a valid torrent_info

**Comment by Falcosc** [comment]
It does not crash in our test client. 
And head of RC_1_0 test client https://github.com/arvidn/libtorrent/blob/RC_1_0/examples/client_test.cpp does return incorrect number of piece hashes in torrent file as well

**Comment by Falcosc** [comment]
I did manually check the
info.length = 2196780193
that are 524 pieces because of info.piece length 4194304
that means we expect 10480 bytes of piecehash but there are 11180 bytes in there

That means all is working fine and the ticket could be closed because it did not crash and the error is correct.

**Comment by sledgehammer999** [comment]
@Falcosc did you see my stackstrace? It will crash if you feed the `torrent_info` into `libtorrent::create_torrent::create_torrent()`.

**Comment by Falcosc** [comment]
The client your add_torrent.

I am not sure if it's ok to direct create it because it may skip some validations.

**Comment by Falcosc** [comment]
@sledgehammer999 plese could you tell how you create the torrent info?

**Comment by Falcosc** [comment]
I did create a test which create a torrent_info and the validation got triggered in RC_1_0

@sledgehammer999 
please could you add the style how you create a torrent as test like this?
https://github.com/arvidn/libtorrent/pull/1531/files

**Comment by arvidn** [comment]
my bet is that the torrent that's being saved is a magnet link that hasn't completed downloading the metadata yet. during that time, the number of pieces is estimated based on the length of bitfield messages. However, those only have 8-pieces granularity. 

Does that theory sound like it fit the failure case?

**Comment by arvidn** [comment]
oh, reading through a bit more. I think I get the problem now. you're *adding* a torrent. you do that by loading it and then saving it (using create_torrent()) in a new place, say your application settings directory. Is that right?
and we would expect torrent_info to report an error on load, _and_ we would expect create_torrent() to not crash.
still though, it seems like unnecessary work to save the torrent via create_torrent, instead of just copying the file.

**Comment by sledgehammer999** [comment]
Sorry for the delay.

Answering arvid's last post: We wrap the torrent_info class in our convenience class.
We load the torrent using the torrent_info constructor that takes an error_code. Then after cosntruction, we call our wrapper's isValid() function which does this
```c++
 return (m_nativeInfo && m_nativeInfo->is_valid() && (m_nativeInfo->num_files() > 0));
```

In this case `torrent_info::is_valid()` returns true on RC_1_0 and false on RC_1_1.  And because it fails on RC_1_1 we are able to abort early and print the error message from error_code.

After we add the torrent we want to save a copy of it for starting up the client. We choose to do this by creating the .torrent file using create_torrent() and the torrent_info that we have in memory. We do this to bypass potential file permissions problems when copying the ondisk original file.
Even if it isn't an approved workflow, the matter is that create_torrent() crashes and it shouldn't. And more specifically torrent_info on RC_1_0 should report as non-valid like RC_1_1 does.
**Unless**, libtorrent is able to "correct" the malformation on the fly. Then RC_1_1 should get an upgrade.

**Comment by arvidn** [comment]
do you have an example torrent file that I could load in RC_1_0 to reproduce this?

**Comment by sledgehammer999** [comment]
Yes see here: https://github.com/qbittorrent/qBittorrent/issues/5480#issuecomment-230411200

**Comment by arvidn** [comment]
testing that torrent in RC_1_0 does result in an error from ``torrent_info`` constructor. As far as I can tell, the line of code that checks the length of the piece hash string is from 2009.

testing the 1.0 tag (``libtorrent-1_0``) also fails with the correct error (this tag was cut 2014).

btw, I'm testing with the ``dump_torrent`` example, which basically just loads a torrent by passing it to the ``torrent_info`` constructor. I wonder if there's some other path to load the torrent that I'm missing.

**Comment by sledgehammer999** [comment]
but `torrent_info::is_valid()` returns true instead of false, right?

**Comment by Falcosc** [comment]
Do you know all way which are used by qbittorend for creating a torrent_info instance?
Because all constructors have the validation. There are some ways to skip the validation which could case the issue of the ticket:

- Constuct with a existing torrent info object
- Clone a existing torrent info and apply some new data to it

And there is an unknown way: use one of the 4 other constructors with strange data which case a skip. I could they anything about this unknown case because I didn't check the constructors for possible ways to skip the validation. It would be more easy if somebody could tell us which constructor was used or at least which of the constructors are implemented in qbittorent.

**Comment by arvidn** [comment]
@sledgehammer999 I see. ``dump_torrent`` looks at the error code and probably terminates, and doesn't test ``is_valid()``. I will comment that check out

**Comment by arvidn** [comment]
here's a fix for 1.0 https://github.com/arvidn/libtorrent/pull/1579

**Comment by arvidn** [comment]
and fix for 1.1 here: https://github.com/arvidn/libtorrent/pull/1580
